walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,76 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;	
 import java.util.Collection;
+import java.util.Collections;	
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if(request.getOptionalAttendees().size() == 0) {
+        return getOpenSlots(events, request);	
+    }	
+
+    Collection<String> allAttendees = new ArrayList<String>();	
+    allAttendees.addAll(request.getAttendees());	
+    allAttendees.addAll(request.getOptionalAttendees());	
+    MeetingRequest allAttendeesRequest = new MeetingRequest(allAttendees, request.getDuration());	
+    Collection<TimeRange> openSlotsForMandatoryAttendees = getOpenSlots(events, request);	
+    Collection<TimeRange> openSlotsForAllAttendees = getOpenSlots(events, new MeetingRequest(allAttendees, request.getDuration()));	
+
+    if(openSlotsForAllAttendees.size() == 0 && request.getAttendees().size() > 0) {	
+      return openSlotsForMandatoryAttendees;	
+    } else {	
+      return openSlotsForAllAttendees;	
+    }	
+  }	
+
+  private Collection<TimeRange> getOpenSlots(Collection<Event> events, MeetingRequest request) {	
+    List<TimeRange> openSlots = new ArrayList<TimeRange>();	
+    int startTime = TimeRange.START_OF_DAY;	
+    int endTime = TimeRange.START_OF_DAY;	
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {	
+      return openSlots;	
+    }	
+
+    // Going through the events and adding an open slot between events for mandatory attendees	
+    for (Event e : events) {	
```

Comment:
Should we sort the events before we iterate and check if it can be added to openSlots?

Reply:
In the instructions it was said that the events are sorted by starting time